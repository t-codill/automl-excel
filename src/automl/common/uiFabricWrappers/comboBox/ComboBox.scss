
.combo-box {
  // font-size: 14px;
  .error-placeholder {
    visibility: hidden;
    margin-top: 5px;
  }

  .ms-ComboBox-CaretDown-button {
    display: flex; 
    padding-left: 0;  
    font-size: 12px;
    .ms-Button-icon{
        height: 23px;
        margin-left: 11px;
    }    
  }

  &.as-search-box {
    .ms-ComboBox-CaretDown-button {
      display: none;
    }
  }
}
.ms-ComboBox-option {
  .ms-Button-flexContainer > span {
    display: inline-block;
    width: 100%;
  }

  .match {
    background-color: white;
  }
}