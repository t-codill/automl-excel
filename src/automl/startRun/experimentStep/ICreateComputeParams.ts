export interface ICreateComputeParams {
    name: string;
    vmSize: string;
    minNodeCount: string;
    maxNodeCount: string;
}
