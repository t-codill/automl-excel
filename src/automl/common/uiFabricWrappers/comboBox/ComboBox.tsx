import * as _ from "lodash";
import * as React from "react";

import { ComboBox as FabricComboBox, DirectionalHint, IComboBox, IComboBoxOption, IComboBoxOptionStyles, IIconProps, SelectableOptionMenuItemType } from "office-ui-fabric-react";
import "./ComboBox.scss";

export interface IProps {
  allowFreeform?: boolean;
  buttonIconProps?: IIconProps;
  className?: string;
  disabled?: boolean;
  label?: string;
  loading?: boolean;
  text?: string;
  placeholder?: string;
  options: IComboBoxOption[];
  initialSelectedKey?: string;
  asSearchBox?: boolean;
  focusOnInitially?: boolean;
  id?: string;
  required?: boolean;
  errorMessage?: string;
  comboBoxOptionStyles?: Partial<IComboBoxOptionStyles>;
  onOptionSelection(selectedKey?: string): void;
  onBlur?(): void;
  onClick?(event: React.MouseEvent<IComboBox, MouseEvent>): void;
}

interface IState {
  options: IComboBoxOption[];
  selectedKey?: string;
}

export class ComboBox extends React.Component<IProps, IState> {
  private readonly comboBoxCalloutId: string;
  private readonly comboBoxRef: React.RefObject<IComboBox> = React.createRef<IComboBox>();

  constructor(props: IProps) {
    super(props);
    this.comboBoxCalloutId = `combobox-callout-${_.uniqueId()}`;

    this.state = {
      options: this.props.options,
      selectedKey: this.props.initialSelectedKey,
    };
  }

  public componentDidMount(): void {
    if (this.props.focusOnInitially) {
      this.openComboBox();
    }
  }

  public componentWillReceiveProps(nextProps: IProps): void {
    if (nextProps.options !== this.props.options) {
      this.setState({
        options: nextProps.options
      });
    }
  }

  public render(): React.ReactNode {

    const placeholder = this.props.loading ? "Loading..." : (this.props.placeholder || "Select or search by name");
    const cssClass = this.props.asSearchBox ? "as-search-box" : "";
    const selectedKey = this.state.selectedKey || this.props.initialSelectedKey;

    return (
      <div className={`combo-box user-input-box ${cssClass}`} id={`${this.props.id || ""}`}>
        <FabricComboBox
          comboBoxOptionStyles={this.props.comboBoxOptionStyles}
          label={this.props.label}
          buttonIconProps={this.props.buttonIconProps}
          className={this.props.className}
          disabled={this.props.disabled}
          placeholder={placeholder}
          options={this.state.options}
          selectedKey={selectedKey}
          text={this.props.text}
          onChange={this.onOptionSelection}
          onPendingValueChanged={_.debounce(this.onValueChange, 100)}
          errorMessage={this.props.errorMessage}
          allowFreeform={true}
          autoComplete={this.props.allowFreeform ? "off" : "on"}
          useComboBoxAsMenuWidth={true}
          onClick={this.handleOnClick}
          onFocus={this.openComboBox}
          onMenuDismissed={this.props.onBlur}
          required={this.props.required}
          calloutProps={{ id: this.comboBoxCalloutId, directionalHint: DirectionalHint.bottomLeftEdge, directionalHintFixed: true }}
          componentRef={this.comboBoxRef}
          styles={{ optionsContainerWrapper: { maxHeight: 400 } }}
        />
      </div>
    );
  }

  private readonly handleOnClick = (event: React.MouseEvent<IComboBox, MouseEvent>) => {
    if (this.props.onClick) {
      this.props.onClick(event);
    }
    this.openComboBox();
  }

  private readonly onOptionSelection = (_e: React.FormEvent<IComboBox>, selectedOption?: IComboBoxOption, _index?: number, typedText?: string): void => {
    let selectedKey: string | undefined;

    if (selectedOption && !selectedOption.disabled) {
      selectedKey = selectedOption.key as string;
    } else if (typedText) {
      if (this.props.allowFreeform) {
        selectedKey = typedText.trim();
      } else if (this.state.options.length > 0) {
        const matchOption = this.state.options.find((option) =>
          !option.disabled &&
          option.itemType !== SelectableOptionMenuItemType.Header &&
          option.itemType !== SelectableOptionMenuItemType.Divider
        );
        if (matchOption) {
          selectedKey = matchOption.key as string;
        }
      }
    }

    this.setState(
      { selectedKey, options: this.props.options },
      () => { this.props.onOptionSelection(selectedKey); }
    );
  }

  private readonly onValueChange = (_option?: IComboBoxOption, _index?: number, typedText?: string): void => {
    if (typedText === undefined) { return; }
    this.openComboBox();

    const regEx = new RegExp(`(.*?)(${_.escapeRegExp(typedText)})(.*)`, "i");
    let newOptions = [];

    for (const option of this.props.options) {
      if (option.itemType === SelectableOptionMenuItemType.Header) {
        newOptions.push(option);
        continue;
      }

      const match = option.text.match(regEx);
      if (match) {
        const newOption = { ...option };
        newOption.data = `<span>${match[1] || ""}<span class="match">${match[2]}</span>${match[3] || ""}</span>`;
        newOptions.push(newOption);
      }
    }

    // Remove headers with no selected options beneath them
    newOptions = newOptions.filter((option, index, _newOptions) => {
      const nextOption = _newOptions[index + 1];

      return option.itemType !== SelectableOptionMenuItemType.Header || (
        option.itemType === SelectableOptionMenuItemType.Header &&
        nextOption && nextOption.itemType !== SelectableOptionMenuItemType.Header
      );
    });

    this.setState({ options: newOptions });
  }

  private readonly openComboBox = (): void => {
    if (this.comboBoxRef.current) {
      this.comboBoxRef.current.focus(true);
    }
  }
}
