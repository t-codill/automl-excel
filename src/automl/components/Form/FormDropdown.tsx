import * as React from "react";
import { Omit } from "../../common/Omit";
import { ComboBox, IProps } from "../../common/uiFabricWrappers/comboBox/ComboBox";
import { FormBaseInput } from "./FormBaseInput";

export class FormDropdown<
    TData extends { [key in TField]: string | undefined },
    TField extends keyof TData
    > extends FormBaseInput<TData, TField, string | undefined, Omit<IProps, "allowFreeform" | "onOptionSelection">>{
    public render(): React.ReactNode {
        return <ComboBox
            allowFreeform={false}
            onOptionSelection={this.onOptionsSelection}
            {...this.props}
            // These props cannot be overridden
            errorMessage={this.state.errorMessage}
            initialSelectedKey={this.state.value}
            comboBoxOptionStyles={{
                root:
                {
                    paddingTop: 0,
                    paddingBottom: 0,
                    minHeight: 20
                }
            }}
        />;
    }

    private readonly onOptionsSelection = (selected?: string) => {
        if (selected) {
            this.setValue(selected);
        }
    }
}
