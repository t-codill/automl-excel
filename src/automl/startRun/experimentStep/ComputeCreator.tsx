import { AzureMachineLearningWorkspacesModels } from "@azure/arm-machinelearningservices";
import { DefaultButton, Panel, PanelType, PrimaryButton } from "office-ui-fabric-react";
import * as React from "react";
import { Accordion } from "../../components/Accordion/Accordion";
import { BaseComponent } from "../../components/Base/BaseComponent";
import { Form } from "../../components/Form/Form";
import { FormTextInput } from "../../components/Form/FormTextInput";
import { Validators } from "../../components/Form/Validators";
import { PopupProgressIndicator } from "../../components/Progress/PopupProgressIndicator";
import { IVMSizeResult, WorkSpaceService } from "../../services/WorkSpaceService";
import { ICreateComputeParams } from "./ICreateComputeParams";
import { MaxNodeInput } from "./MaxNodeInput";
import { MinNodeInput } from "./MinNodeInput";
import { VmSizeSelector } from "./VmSizeSelector";

export interface IComputeCreatorProps {
    vmSizes: IVMSizeResult | undefined;
    onComputeCreated(compute: AzureMachineLearningWorkspacesModels.ComputeResource): void;
    onCancel(): void;
}

export interface IComputeCreatorState {
    creating: boolean;
    minNodeCount: string;
}

export class ComputeCreator extends BaseComponent<IComputeCreatorProps, IComputeCreatorState,
    {
        workSpaceService: WorkSpaceService;
    }> {
    protected readonly serviceConstructors = {
        workSpaceService: WorkSpaceService
    };

    protected readonly getData = undefined;

    constructor(props: IComputeCreatorProps) {
        super(props);
        this.state = {
            creating: false,
            minNodeCount: "0"
        };
    }

    public render(): React.ReactNode {
        return <>
            {!this.state.creating &&
                <Panel
                    isOpen={true}
                    type={PanelType.medium}
                    onDismiss={this.cancel}
                    isHiddenOnDismiss={true}
                    headerText="Create a New Compute"
                    closeButtonAriaLabel="Close"
                >
                    <Form onSubmit={this.createCompute} onUpdated={this.onFormUpdated}>
                        <FormTextInput<ICreateComputeParams, "name">
                            field="name"
                            required={true}
                            label="Compute name"
                            placeholder="Compute name"
                            ariaLabel="Please enter compute name here"
                            validators={[
                                Validators.required("Compute name is required"),
                                Validators.regex(/^[a-zA-Z][a-zA-Z0-9-]{1,15}$/,
                                    "Compute name should start with a letter, be between 2 and 16 character, and only include letters (a-zA-Z), numbers (0-9) and '-'")]}
                        />
                        <VmSizeSelector vmSizes={this.props.vmSizes} />
                        <Accordion
                            exclusive={true}
                            items={[
                                {
                                    title: "Additional Settings",
                                    collapsed: true,
                                    disabled: false,
                                    element: <>
                                        <MinNodeInput />
                                        <MaxNodeInput minNodeCount={this.state.minNodeCount} />
                                    </>
                                }
                            ]} />
                        <div className="form-footer">
                            <DefaultButton text="Cancel" onClick={this.cancel} />
                            <PrimaryButton text="Create" type="submit" />
                        </div>
                    </Form>
                </Panel>
            }
            {this.state.creating &&
                <PopupProgressIndicator
                    title="Creating a new compute..."
                    description="Creating a new compute. This may take a couple minutes."
                />}
        </>;
    }

    private readonly onFormUpdated = (key: keyof ICreateComputeParams, value: ICreateComputeParams[keyof ICreateComputeParams]) => {
        if (!value) {
            return;
        }
        if (key === "minNodeCount") {
            this.setState({ minNodeCount: value });
        }
    }

    private readonly cancel = () => {
        this.props.onCancel();
    }

    private readonly createCompute = async (data: ICreateComputeParams) => {
        if (!data
            || !data.name
            || !data.vmSize
            || !data.minNodeCount
            || !data.maxNodeCount) {
            return;
        }

        this.setState({
            creating: true
        });
        this.logUserAction("CreateCompute", { data: JSON.stringify(data) });
        const compute = await this.services.workSpaceService.createCompute(
            data.name,
            data.vmSize,
            parseInt(data.minNodeCount, 10),
            parseInt(data.maxNodeCount, 10)
        );

        this.setState({ creating: false });

        if (!compute) {
            return;
        }

        this.props.onComputeCreated(compute);
    }
}
