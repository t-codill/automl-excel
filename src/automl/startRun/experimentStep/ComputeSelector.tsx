import { AzureMachineLearningWorkspacesModels } from "@azure/arm-machinelearningservices";
import { reduce } from "lodash";
import { Icon, IDropdownOption, Link } from "office-ui-fabric-react";
import * as React from "react";
import { computeUtils } from "../../common/utils/computeUtils";
import { FormDropdown } from "../../components/Form/FormDropdown";
import { Validators } from "../../components/Form/Validators";
import { LabelWithCallout } from "../../components/Label/LabelWithCallout";
import { IExperimentStepCallback } from "./ExperimentStep";

interface IComputeSelectorProps {
    computes: AzureMachineLearningWorkspacesModels.ComputeResource[] | undefined;
    selectedComputeId: string | undefined;
    readonly: boolean;
    refreshing: boolean;
    onEditClick(): void;
}

export class ComputeSelector extends React.PureComponent<IComputeSelectorProps> {

    public render(): React.ReactNode {
        const dropdownId = "experimentStepComputeSelector";
        return <>
            <LabelWithCallout required={true} htmlFor={dropdownId} labelText="Select a training compute">
                A cloud compute resource is required to run the experiment in Azure.
                You can use your existing Azure ML compute resource, or create a new one.
                Note that compute resource has its own costs, refer to the &nbsp;
                <Link className="ms-CalloutExample-link" href="https://azure.microsoft.com/en-us/pricing/details/machine-learning-service/" target="_blank">
                    pricing details <Icon iconName="NavigateExternalInline" />
                </Link> for more information.
                Computes marked "profiling enabled" support additional data profiling capabilities.</LabelWithCallout>
            <FormDropdown<IExperimentStepCallback, "computeId">
                field="computeId"
                id={dropdownId}
                required={true}
                placeholder="Select a Compute..."
                options={this.loadOptions()}
                buttonIconProps={
                    {
                        iconName: this.getIcon()
                    }
                }
                onClick={this.props.readonly ? this.props.onEditClick : undefined}
                validators={[
                    Validators.required("Compute is required")
                ]}
            />
        </>;
    }
    private readonly getIcon = () => {
        if (this.props.refreshing) {
            return "ProgressRingDots";
        }
        if (this.props.readonly) {
            return "Edit";
        }
        return "ChevronDown";
    }

    private readonly loadOptions = () => {
        return reduce(this.props.computes,
            (options, compute) => this.computeOptionsReducer(options, compute),
            [] as IDropdownOption[]);
    }

    private readonly computeOptionsReducer = (options: IDropdownOption[], compute: AzureMachineLearningWorkspacesModels.ComputeResource) => {
        const computeUnion = compute.properties;
        if (computeUnion) {
            switch (computeUnion.computeType) {
                case "AmlCompute": {
                    options.push(this.getComputeOptions(compute, computeUtils.isComputeProfilingEnabled(compute)));
                    break;
                }
                case "VirtualMachine": {
                    options.push(this.getComputeOptions(compute, true));
                    break;
                }
                default:
            }
        }
        return options;
    }

    private readonly getComputeOptions = (compute: AzureMachineLearningWorkspacesModels.ComputeResource, profilingEnabled: boolean) => {
        const provisioning = this.getProvisioningInfo(compute);
        const profilingText = this.getProfilingText(profilingEnabled);
        const option: IDropdownOption = {
            key: compute.id || "",
            text: `${compute.name}${provisioning.message}${profilingText}`,
            disabled: !provisioning.isSuccessful
        };
        return option;
    }

    private readonly getProfilingText = (profilingEnabled: boolean) => {
        return profilingEnabled ? " (profiling enabled)" : "";
    }

    private readonly getProvisioningInfo = (compute: AzureMachineLearningWorkspacesModels.ComputeResource) => {
        const isComputeUsable = computeUtils.isComputeUsable(compute);
        const isProvisioningSuccessful = compute.properties && compute.properties.provisioningState === "Succeeded";
        let message = "";
        if (!isProvisioningSuccessful) {
            message = ` (${(compute.properties && compute.properties.provisioningState) ? compute.properties.provisioningState : "Unknown"})`;
        } else if (!isComputeUsable) {
            message = " (Unusable)";
        }
        return { message, isSuccessful: isProvisioningSuccessful && isComputeUsable };
    }
}
