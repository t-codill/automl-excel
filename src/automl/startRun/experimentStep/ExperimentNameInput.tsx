import * as React from "react";
import { FormComboBox } from "../../components/Form/FormComboBox";
import { Validators } from "../../components/Form/Validators";
import { IExperimentStepCallback } from "./ExperimentStep";

interface IExperimentNameInputProps {
    experimentNames: string[] | undefined;
    experimentName: string | undefined;
    readonly: boolean;
    onEditClick(): void;
}

export class ExperimentNameInput extends React.PureComponent<IExperimentNameInputProps> {
    public render(): React.ReactNode {
        const comboBoxId = "experimentNameInput";
        return <>
            <FormComboBox<IExperimentStepCallback, "experimentName">
                field="experimentName"
                required={true}
                placeholder="Enter experiment name or select"
                options={this.loadOptions()}
                onClick={this.onEditClick}
                label="Experiment name"
                id={comboBoxId}
                buttonIconProps={{ iconName: this.props.readonly ? "Edit" : "ChevronDown" }}
                validators={[
                    Validators.required("Experiment name is required"),
                    Validators.regex(/^[a-zA-Z0-9][\w-]{2,35}$/,
                        "Experiment name must be 3-36 characters, start with a letter or a number, and can only contain letters, numbers, underscores, and dashes.")
                ]}
            />
        </>;
    }

    private readonly onEditClick = (): void => {
        this.props.onEditClick();
    }

    private readonly loadOptions = () => {
        if (!this.props.experimentNames) {
            return [];
        }

        return this.props.experimentNames.map((name) => ({ key: name, text: name }));
    }
}
