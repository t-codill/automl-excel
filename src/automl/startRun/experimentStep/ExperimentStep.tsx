import { AzureMachineLearningWorkspacesModels } from "@azure/arm-machinelearningservices";
import { map, uniq } from "lodash";
import { DefaultButton, Dialog, DialogFooter, DialogType, Icon, Link, PrimaryButton } from "office-ui-fabric-react";
import * as React from "react";
import { BaseComponent } from "../../components/Base/BaseComponent";
import { Form } from "../../components/Form/Form";
import { RunHistoryService } from "../../services/RunHistoryService";
import { IVMSizeResult, WorkSpaceService } from "../../services/WorkSpaceService";
import { ComputeCreator } from "./ComputeCreator";
import { ComputeSelector } from "./ComputeSelector";
import { ExperimentNameInput } from "./ExperimentNameInput";

export interface IExperimentStepProps {
    experimentName: string | undefined;
    compute: AzureMachineLearningWorkspacesModels.ComputeResource | undefined;
    readOnly: boolean;
    onEditClick(): void;
    onCancel(): void;
    onNext(
        experimentName: string | undefined,
        compute: AzureMachineLearningWorkspacesModels.ComputeResource | undefined
    ): void;
}
export interface IExperimentStepCallback {
    experimentName: string | undefined;
    computeId: string | undefined;
}

export interface IExperimentStepState {
    experimentNames: string[] | undefined;
    computes: AzureMachineLearningWorkspacesModels.ComputeResource[] | undefined;
    selectedCompute: AzureMachineLearningWorkspacesModels.ComputeResource | undefined;
    showComputeCreator: boolean;
    vmSizes: IVMSizeResult | undefined;
    showCancelDialog: boolean;
    computeRefreshing: boolean;
}

export class ExperimentStep
    extends BaseComponent<IExperimentStepProps, IExperimentStepState, {
        workSpaceService: WorkSpaceService;
        runHistoryService: RunHistoryService;
    }> {
    protected readonly serviceConstructors = {
        workSpaceService: WorkSpaceService,
        runHistoryService: RunHistoryService
    };

    constructor(props: IExperimentStepProps) {
        super(props);
        this.state = {
            experimentNames: undefined,
            computes: undefined,
            selectedCompute: props.compute,
            showComputeCreator: false,
            vmSizes: undefined,
            showCancelDialog: false,
            computeRefreshing: false
        };
    }

    public render(): React.ReactNode {
        return <>
            <Form
                onSubmit={this.onSubmit}>
                <div className="start-run-step-content">
                    <div className="ms-md12 ms-lg8 ms-xl6 ms-xxl4">
                        <ExperimentNameInput
                            experimentName={this.props.experimentName}
                            readonly={this.props.readOnly}
                            onEditClick={this.props.onEditClick}
                            experimentNames={this.state.experimentNames}
                        />
                    </div>
                    <div className="ms-md12 ms-lg8 ms-xl6 ms-xxl4">
                        <ComputeSelector
                            computes={this.state.computes}
                            selectedComputeId={this.state.selectedCompute && this.state.selectedCompute.id}
                            readonly={this.props.readOnly}
                            refreshing={this.state.computeRefreshing}
                            onEditClick={this.props.onEditClick}
                        />
                        {!this.props.readOnly &&
                            <Link
                                id="linkCreateCompute"
                                onClick={this.onCreateComputeClick}>
                                <Icon iconName="Devices3" style={{ marginRight: 5 }} />
                                <span>Create a new compute</span>
                            </Link>
                        }
                        {!this.props.readOnly &&
                            <Link
                                id="linkRefreshCompute"
                                onClick={this.onRefreshComputeClick} >
                                <Icon iconName="Refresh" style={{ marginRight: 5, marginLeft: 15 }} />
                                <span>Refresh Compute</span>
                            </Link>
                        }
                    </div>
                </div>

                {!this.props.readOnly &&
                    <div className="form-footer">
                        <DefaultButton text="Cancel" onClick={this.showCancelDialog} />
                        <PrimaryButton type="submit" text="Next" />
                    </div>
                }
            </Form>

            {this.state.showComputeCreator &&
                <ComputeCreator
                    {...this.props}
                    {...this.state}
                    onComputeCreated={this.onComputeCreated}
                    onCancel={this.onComputeCreateCancel}
                />
            }
            <Dialog
                hidden={!this.state.showCancelDialog}
                onDismiss={this.closeCancelDialog}
                dialogContentProps={{
                    type: DialogType.normal,
                    title: "Cancel New Experiment?",
                    subText: `Are you sure you want to cancel creating a new Experiment?`
                }}
                modalProps={{
                    titleAriaId: "cancelDialogLabel",
                    subtitleAriaId: "cancelSubtextLabel",
                    isBlocking: false,
                    containerClassName: "ms-dialogMainOverride"
                }}
            >
                <DialogFooter>
                    <PrimaryButton onClick={this.cancelRun} text="Yes" />
                    <DefaultButton onClick={this.closeCancelDialog} text="No" />
                </DialogFooter>
            </Dialog>
        </>;
    }

    protected readonly getData = async () => {
        this.getExperimentNames();
        this.getComputes();
        this.getVmSizes();
    }

    private readonly onCreateComputeClick = () => {
        this.setState({ showComputeCreator: true });
    }

    private readonly onRefreshComputeClick = () => {
        this.getComputes();
    }

    private readonly onComputeCreated = (compute: AzureMachineLearningWorkspacesModels.ComputeResource) => {
        this.setState({ showComputeCreator: false, selectedCompute: compute });
        this.getComputes();
    }

    private readonly onComputeCreateCancel = () => {
        this.setState({ showComputeCreator: false });
    }

    private readonly getExperimentNames = async (): Promise<void> => {
        const experiments = await this.services.runHistoryService.listExperiments();
        if (!experiments) {
            return;
        }
        const experimentNames = uniq(map(experiments, (experiment) => experiment.name || ""));
        this.setState({ experimentNames });
    }

    private readonly getComputes = async (): Promise<void> => {
        this.setState({ computeRefreshing: true });
        const computes = await this.services.workSpaceService.listComputes();
        if (!computes) {
            return;
        }
        this.setState({ computes, computeRefreshing: false });
    }

    private readonly getVmSizes = async (): Promise<void> => {
        const vmSizes = await this.services.workSpaceService.listVmSizes();
        if (!vmSizes) {
            return;
        }
        this.setState({ vmSizes });
    }

    private readonly onSubmit = (values: IExperimentStepCallback) => {
        if (!values.computeId || !this.state.computes) {
            return;
        }
        this.props.onNext(values.experimentName, this.state.computes.find((c) => c.id === values.computeId));
    }

    private readonly showCancelDialog = (): void => {
        this.setState({ showCancelDialog: true });
    }

    private readonly closeCancelDialog = (): void => {
        this.setState({ showCancelDialog: false });
    }

    private readonly cancelRun = () => {
        this.props.onCancel();
        this.closeCancelDialog();
    }

}
