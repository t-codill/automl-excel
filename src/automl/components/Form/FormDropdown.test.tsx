import { shallow, ShallowWrapper } from "enzyme";
import { IComboBoxOption } from "office-ui-fabric-react";
import * as React from "react";
import { ComboBox, IProps } from "../../common/uiFabricWrappers/comboBox/ComboBox";
import { IExperimentStepCallback } from "../../startRun/experimentStep/ExperimentStep";
import { FormDropdown } from "./FormDropdown";

jest.mock("./FormBaseInput");

describe("FormDropdown", () => {
    let wrapper: ShallowWrapper<{}, { value: string | undefined }>;
    let formDropdownProps: IProps;
    beforeEach(() => {
        const options: IComboBoxOption[] = [{ key: "key1", text: "text1" }, { key: "key2", text: "text2" }, { key: "key3", text: "text3" }];
        wrapper = shallow(
            <FormDropdown<IExperimentStepCallback, "experimentName">
                field="experimentName"
                required={true}
                placeholder="test place holder"
                options={options}
                label="test label"
                id="DropdownId"
            />);
        formDropdownProps = wrapper.find(ComboBox)
            .props();
    });

    it("should render", () => {
        expect(wrapper)
            .toMatchSnapshot();
    });

    it("should on change", () => {
        if (formDropdownProps.onOptionSelection) {
            formDropdownProps.onOptionSelection("test text");
        }
        expect(wrapper.state("value"))
            .toBe("test text");
    });

    it("should handle no selection", () => {
        wrapper.setProps({ enableSearch: true });
        if (formDropdownProps.onOptionSelection) {
            formDropdownProps.onOptionSelection(undefined);
        }
        expect(wrapper.state("value"))
            .toBe(undefined);
    });
});
