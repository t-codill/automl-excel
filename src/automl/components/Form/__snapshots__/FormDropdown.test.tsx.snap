// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormDropdown should render 1`] = `
<ComboBox
  allowFreeform={false}
  comboBoxOptionStyles={
    Object {
      "root": Object {
        "minHeight": 20,
        "paddingBottom": 0,
        "paddingTop": 0,
      },
    }
  }
  field="experimentName"
  id="DropdownId"
  label="test label"
  onOptionSelection={[Function]}
  options={
    Array [
      Object {
        "key": "key1",
        "text": "text1",
      },
      Object {
        "key": "key2",
        "text": "text2",
      },
      Object {
        "key": "key3",
        "text": "text3",
      },
    ]
  }
  placeholder="test place holder"
  required={true}
/>
`;
